NimBLEDevice::get/setPower support full power range.

thekurtovic opened the pull request:
* Revert NimBLEDevice::setPower to take esp_power_level_t.
* * Fixes examples which set tx power.
* * Avoids limiting valid power levels, using the datatype ensures valid input option.
* * Since preprocessor was used inside function, separate implementations for ESP and non-ESP platforms at compile-time.
* Added NimBLEDevice::getPowerLevel for ESP platform to get esp_power_level_t return value.
* NimBLEDevice::getPower support full power range based on IDF version.

h2zero commented:
This is nice but I was trying to remove as much espressif specific functions/types as I could so it's seamless with other devices/oses.

thekurtovic commented:
I see. What about just making the functions do nothing for ESP and emit a warning for the application to use `esp_ble_tx_power_set`/`esp_ble_tx_power_get` directly?

h2zero commented:
That's also a possibility but I think the controller command might work with the newer esp32 chips, need to test.

thekurtovic commented:
Closing due to excessive platform specific handling.

h2zero commented:
Too soon, I will do something with this :)

h2zero commented:
I think this could work, the only request I would make is to have the values specified in dbm, which is what NimBLE natively works with. Just need to remove the `esp_power_level_t` references and add conversions from dbm back and forth to `esp_power_level_t`.

h2zero commented:
@thekurtovic I have added a commit and rebased to master, please take a look when you can.

thekurtovic commented:
Would something like this make sense for the repo? If other platforms are supported in the future which require specific handling for power this could get ugly. This cleans up the code within `NimBLEDevice` regarding TX power, and moves the platform specific stuff into their own headers. Created a single function for handling ESP32 conversion, which should be easier to maintain.
Not sure if it would require another option within the menuconfig to select the platform.
https://github.com/h2zero/esp-nimble-cpp/commit/6b90d6cff256f69359ccec56b6abaf891ec74d0f



> @thekurtovic I have added a commit and rebased to master, please take a look when you can.

It doesn't fulfil the goal of the PR. There are edge cases where some chips do not follow the pattern of incrementing by 3. Most notably it appears that going forward all chips max out at 20 dbm, while 21 has become deprecated.

h2zero commented on src/NimBLEDevice.cpp:
```diff
@@ -481,31 +478,21 @@ bool NimBLEDevice::setPower(int8_t dbm) {
+    return -127; // CONFIG_IDF_TARGET_ESP32P4 does not support esp_ble_tx_power_get
 #  endif
 
+    int pwr = esp_ble_tx_power_get(ESP_BLE_PWR_TYPE_DEFAULT);
```
In theory the current power level should never be invalid I though think, also `ESP_PWR_LVL_INVALID` isn't defined for all chips/idf versions.

h2zero commented:
> Edit: Wait I think I misunderstood what the code was doing. The 20 vs 21 problem should be avoided since it would cast into the same value. Also the ESP32-H2 enums get mapped into a shared values so they should be okay too.

We could add a check for this, 20dbm is actually the max allowed by BLE specs so that's safe 😄

thekurtovic commented:
Okay spent a bit more time looking at the changes, I didn't notice you added methods for setting power level for esp32.
The calculation for dbm to power level will not give the expected value under some combinations of chip and IDF version, but overall it will cover almost all cases and the user can always just set the power level using the enum instead.
LGTM, tested on esp32-s3.

This isn't related to the PR but are you able to get a valid reading on an esp32's power level? I get 0xFF anytime the value is read, so I can't properly test the `getPower` methods. There's no error when setting the power so I assume it's fine. The last time I remember being able to read the power level was in IDF 4.4.x, currently it doesn't matter if I check at initialization or when a client is connected, always get 0xFF.

h2zero commented:
And on the C6:
```
E (469) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -30
E (479) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -29
E (489) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -28
E (489) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -27
E (499) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -26
E (509) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -25
E (519) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -24
E (529) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -23
E (539) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -22
E (539) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -21
E (549) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -20
E (559) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -19
E (569) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 9, requested -18
Power level: -15, requested -17
Power level: -15, requested -16
Power level: -15, requested -15
Power level: -12, requested -14
Power level: -12, requested -13
Power level: -12, requested -12
Power level: -9, requested -11
Power level: -9, requested -10
Power level: -9, requested -9
Power level: -6, requested -8
Power level: -6, requested -7
Power level: -6, requested -6
Power level: -3, requested -5
Power level: -3, requested -4
Power level: -3, requested -3
Power level: 0, requested -2
Power level: 0, requested -1
Power level: 0, requested 0
Power level: 0, requested 1
Power level: 3, requested 2
Power level: 3, requested 3
Power level: 3, requested 4
Power level: 6, requested 5
Power level: 6, requested 6
Power level: 6, requested 7
Power level: 9, requested 8
Power level: 9, requested 9
Power level: 9, requested 10
Power level: 12, requested 11
Power level: 12, requested 12
Power level: 12, requested 13
Power level: 15, requested 14
Power level: 15, requested 15
Power level: 15, requested 16
Power level: 18, requested 17
Power level: 18, requested 18
Power level: 18, requested 19
Power level: 21, requested 20
Power level: 21, requested 21
Power level: 21, requested 22
E (679) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 21, requested 23
E (689) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 21, requested 24
E (699) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 21, requested 25
E (709) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 21, requested 26
E (709) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 21, requested 27
E (719) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 21, requested 28
E (729) NimBLEDevice: esp_ble_tx_power_set: rc=-1
Power level: 21, requested 29
```

Seems like all the devices start at 9dbm, interesting that the C6 actually goes to 21dbm and not capped at 20.

h2zero commented:
I've added the `powerType` parameters to `set/getPowerLevel` just in case anyone wants to use them.

h2zero commented:
> This isn't related to the PR but are you able to get a valid reading on an esp32's power level? I get 0xFF anytime the value is read, so I can't properly test the getPower methods. There's no error when setting the power so I assume it's fine. The last time I remember being able to read the power level was in IDF 4.4.x, currently it doesn't matter if I check at initialization or when a client is connected, always get 0xFF.

Looks like the is a bug with the esp32s3, seems to work well with other versions.

h2zero commented:
Aaaand just like that I found the problem with the esp32s3 right after merging this lol.

@thekurtovic if we call `esp_ble_tx_power_get_enhanced(ESP_BLE_ENHANCED_PWR_TYPE_DEFAULT, 0);` instead of `esp_ble_tx_power_get(ESP_BLE_PWR_TYPE_DEFAULT);` the esp32s3 will return the expected value.

thekurtovic commented:
> I'm curious if you have any examples? I know there are a couple settings out there that were not multiples of 3 but they are deprecated.

20 dbm was the main concern, but I see you've addressed it. Other than that I think only esp32-h2 could get an incorrect value if on IDF 5.1.0-5.1.1 and trying to use 16,17, or 19; though the values are now deprecated and round up to 18 and 20 now.

CONFIG_BT_CTRL_DFT_TX_POWER_LEVEL defaults to P9 so that makes sense regarding the start value.


> if we call `esp_ble_tx_power_get_enhanced(ESP_BLE_ENHANCED_PWR_TYPE_DEFAULT, 0);` instead of `esp_ble_tx_power_get(ESP_BLE_PWR_TYPE_DEFAULT);` the esp32s3 will return the expected value.

Interesting, good find.
I wonder if the enhanced functions can be used in place of the standard ones for all variants of esp32.

h2zero commented:
> I wonder if the enhanced functions can be used in place of the standard ones for all variants of esp32.

Sadly not but I also found that the same result is achieved by calling `esp_ble_tx_power_get(**ESP_BLE_PWR_TYPE_CONN_HDL3**);` As the lower level call gets the same value that way. I will PR the proper fix upstream but this will need to be worked around here regardless.

h2zero commented:
Looks like this was fixed upstream for v5.4. https://github.com/espressif/esp-idf/commit/4108a5cc4cb96039ed861703dda06f7f3a5969c3

h2zero commented:
@thekurtovic What idf version are you on? Looks like the only version with this issue that has been release is 5.2.3. 5.2.2 and 5.3 don't have the "enchanced" function that broke this.

thekurtovic commented:
Thanks for looking into it.
5.1.4.241008, according to version.txt. I tried `esp_ble_tx_power_get_enhanced` out on esp32-s3 and it works for me. 
I'll check if 5.3 is any different.

h2zero commented:
Interesting, I had a look and that commit you're using is not released, the actual 5.1.4 release does not contain the offending code.

thekurtovic commented:
I'll give it a try, though planning on migrating over to 5.3 sometime soon.

I forgot to mention this earlier but `esp_ble_tx_power_get` can supposedly return a negative value. Should a check be added for this and return 0xFF to indicate error?
https://github.com/espressif/esp-idf/blob/master/components/bt/include/esp32c3/include/esp_bt.h#L503-L509
